### PR TITLE
Add support for floating-point comparison

### PR DESCRIPTION
**Pull request by nmostafa**
Add support for fcmp instruction in llvm dialect and EDSC builders.

### CODE REVIEW

File: lib/LLVMIR/IR/LLVMDialect.cpp [hunk 1]
```diff
@@ -73,15 +82,29 @@ static ParseResult parseICmpOp(OpAsmParser *parser, OperationState *result) {
...
+    Optional<ICmpPredicate> predicate;
+    predicate = symbolizeICmpPredicate(predicateStr.getValue());
+    if (!predicate)
     return parser->emitError(predicateLoc)
            << "'" << predicateStr.getValue()
            << "' is an incorrect value of the 'predicate' attribute";
+    predicateValue = static_cast<int64_t>(predicate.getValue());
+  } else {
+    Optional<FCmpPredicate> predicate;
+    predicate = symbolizeFCmpPredicate(predicateStr.getValue());
```
Comment (River707) [inline lib/LLVMIR/IR/LLVMDialect.cpp]: Combine declaration and assignment here.

File: lib/LLVMIR/IR/LLVMDialect.cpp [hunk 2]
```diff
@@ -73,15 +82,29 @@ static ParseResult parseICmpOp(OpAsmParser *parser, OperationState *result) {
...
     return parser->emitError(predicateLoc,
                              "expected 'predicate' attribute of string type");
-  Optional<ICmpPredicate> predicateValue =
-      symbolizeICmpPredicate(predicateStr.getValue());
-  if (!predicateValue)
+
+  int64_t predicateValue = 0;
+  if (std::is_same<CmpPredicateType, ICmpPredicate>()) {
+    Optional<ICmpPredicate> predicate;
+    predicate = symbolizeICmpPredicate(predicateStr.getValue());
```
Comment (River707) [inline lib/LLVMIR/IR/LLVMDialect.cpp]: Combine declaration and assignment here.

File: lib/LLVMIR/IR/LLVMDialect.cpp [hunk 3]
```diff
@@ -73,15 +82,29 @@ static ParseResult parseICmpOp(OpAsmParser *parser, OperationState *result) {
...
+    if (!predicate)
     return parser->emitError(predicateLoc)
```
Comment (River707) [inline lib/LLVMIR/IR/LLVMDialect.cpp]: This looks like it needs to be clang-formatted.

File: test/EDSC/builder-api-test.cpp [hunk 4]
```diff
@@ -474,6 +474,64 @@ TEST_FUNC(select_op) {
...
+    edsc::intrinsics::select(B(i, j) >= B(i+one, j), *A(zero, zero), *A(i, j));
+    edsc::intrinsics::select(B(i, j) <= B(i+one, j), *A(zero, zero), *A(i, j));
+    edsc::intrinsics::select(B(i, j) < B(i+one, j), *A(zero, zero), *A(i, j));
+    edsc::intrinsics::select(B(i, j) > B(i+one, j), *A(zero, zero), *A(i, j));
+  });
+
+  // CHECK-LABEL: @select_op
+  //      CHECK: affine.for %{{.*}} = 0 to 1 {
+  // CHECK-NEXT:   affine.for %{{.*}} = 0 to 1 {
+  //  CHECK-DAG:     {{.*}} = cmpf "oeq"
```
Comment (River707) [inline test/EDSC/builder-api-test.cpp]: Can we drop the '{{.*}} = ' on all of these? It doesn't look related to the test.

File: test/LLVMIR/convert-to-llvmir.mlir [hunk 5]
```diff
@@ -477,3 +477,39 @@ func @cond_br_same_target(%arg0: i1, %arg1: i32, %arg2 : i32) -> (i32) {
 // CHECK:      ^[[dummyBlock]]:
 // CHECK-NEXT:  llvm.br ^[[origBlock]](%arg2 : !llvm.i32)
 }
+
+// CHECK-LABEL: func @fcmp(%arg0: !llvm.float, %arg1: !llvm.float) {
+func @fcmp(f32, f32) -> () {
+^bb0(%arg0: f32, %arg1: f32):
+  // CHECK: {{.*}} = llvm.fcmp "oeq" %arg0, %arg1 : !llvm.float
+  // CHECK-NEXT: {{.*}} = llvm.fcmp "ogt" %arg0, %arg1 : !llvm.float
```
Comment (River707) [inline test/LLVMIR/convert-to-llvmir.mlir]: Same here, drop the '{{.*}} = '

File: test/mlir-cpu-runner/simple.mlir [hunk 6]
```diff
@@ -31,3 +31,21 @@ func @foo(%a : memref<1x1xf32>) -> memref<1x1xf32> {
...
+  %3 = load %a[%c0, %c0] : memref<1x1xf32>
+  %5 = cmpf "olt", %3, %cf0 : f32
+  %6 = select %5, %cf0, %3 : f32
+  store %6, %a[%c0, %c0] : memref<1x1xf32>
+  return %a : memref<1x1xf32>
+}
+// CMPTEST: 1.000000e+03 
+// CMPTEST-NEXT: 1.000000e+03 
+// CMPTEST2: 0.000000e+00
+// CMPTEST2-NEXT: 0.000000e+00
```
Comment (River707) [inline test/mlir-cpu-runner/simple.mlir]: Missing newline on  each of the test files.

File: test/EDSC/builder-api-test.cpp (hunk 4 above)
Comment (nmostafa) [reply test/EDSC/builder-api-test.cpp]: Out of curiosity, why would that matter ?

Comment (River707) [reply test/EDSC/builder-api-test.cpp]: It's a testing principle to not have things unrelated to what you are testing as part of the test:
https://github.com/tensorflow/mlir/blob/master/g3doc/TestingGuide.md

File: test/mlir-cpu-runner/simple.mlir (hunk 6 above)
Comment (joker-eph) [inline test/mlir-cpu-runner/simple.mlir]: Why are you adding a runtime test here? The IR test lowering to LLVM seem to already cover it.

Comment (nmostafa) [reply test/mlir-cpu-runner/simple.mlir]: Are you referring to `convert-to-llvmir.mlir` ? I added this to test lowering to llvm bitcode and actual execution (in case mapping the cmp predicates is faulty).

Comment (joker-eph) [reply test/mlir-cpu-runner/simple.mlir]: Yes, `convert-to-llvmir.mlir` is already testing that the standard dialect to LLVM dialect is correct. 
To check that we correctly convert the LLVM dialect to LLVM bitcode you should add a test in `test/Target/llvmir.mlir`.

The cpu-runner tests are not intended to cover this but intended to exercise the JIT path.

Comment (nmostafa) [reply test/mlir-cpu-runner/simple.mlir]: Ok, will add a test to `llvmir.mlir` and remove the cpu-runner one.

File: lib/Conversion/StandardToLLVM/ConvertStandardToLLVM.cpp [hunk 7]
```diff
@@ -883,7 +884,26 @@ struct CmpIOpLowering : public LLVMLegalizationPattern<CmpIOp> {
     rewriter.replaceOpWithNewOp<LLVM::ICmpOp>(
         op, lowering.convertType(cmpiOp.getResult()->getType()),
         rewriter.getI64IntegerAttr(
-            static_cast<int64_t>(convertCmpIPredicate(cmpiOp.getPredicate()))),
+            static_cast<int64_t>(convertCmpPredicate<LLVM::ICmpPredicate>(cmpiOp.getPredicate()))),
```
Comment (joker-eph) [inline lib/Conversion/StandardToLLVM/ConvertStandardToLLVM.cpp]: I think this is going over 80 cols, did you clang-format your changes?

Comment (nmostafa) [reply lib/Conversion/StandardToLLVM/ConvertStandardToLLVM.cpp]: Fixed.
